tests/test_jsonable_encoder.py

```diff
@@ -88,6 +88,42 @@ def test_encode_dict():
     }
 
 
+def test_encode_dict_with_nonprimative_keys():
+    class CustomString:
+        value: str
+
+        def __init__(self, value: str) -> None:
+            self.value = value
+
+        def __eq__(self, other) -> bool:
+            return isinstance(other, CustomString) and self.value == other.value
+
+        def __hash__(self):
+            return hash(self.value)
+
+    assert jsonable_encoder(
+        {CustomString("foo"): "bar"},
+        custom_encoder={CustomString: lambda v: v.value}
+    ) == {"foo": "bar"}
+
+def test_encode_dict_with_custom_encoder_keys():
+    assert jsonable_encoder(
+        {"foo": "bar"},
+        custom_encoder={str: lambda v: "_" + v}
+    ) == {"_foo": "_bar"}
```

Comment:
This test was added as my original version had a bypass for primative key types in `encode_dict`, this gave a 10~20% the performance bump however is incorrect behaviour as it ignores a custom_encoder. This test was added to ensure that the behaviour is not accidentally broken.